src/monitoring.tbasicu

```diff
@@ -0,0 +1,167 @@
+    
+  end function
+
+  
+  Function Rate()
+  
+    LOCAL x, y, z, distance, difficulty, capacity, counter AS Double
+    Local context as GameContext at me.contextPtr
+    Local totalDistance, totalCapacity as Double
+    
+    tbgl_entityGetPos(uScreenGame.player.scene,uScreenGame.player.entity,x,y,z,%TRUE) 'Absolute Player position
+    
+    for i as integer = 1 to uScreenGame.bubbleDirector.bubbleTotalCount
+      capacity = uScreenGame.bubbleDirector.bubble(i).capacity ' Bubble capacity
+      if capacity = 0 then iterate for
+      incr counter
+      distance = tbgl_entityGetDistancePos(uScreenGame.player.scene,uScreenGame.bubbleDirector.bubble(i).entity,x,y,z) ' Bubble---Player Distance
+      totalDistance += distance
+      totalCapacity += capacity
+      me.oneData.difficulty += POW3(1/capacity) * POW2(distance) ' One far big bubble is easier to catch than many close little ones 
+    next
+    
+    me.oneData.meanFrameRate += context.framerate
+    me.oneData.meandistance += totalDistance / counter
+    me.oneData.meanCapacity += totalCapacity / counter
+    counter = 0 
+  end function
+  
+  Function Monitor()
+'[breakpoint] <Any note here. Breakpoint will be set to line following this statement>
+'[t] on second pass me.contextPtr is lost (reset to 0 , why ? I think it's because storage() was redim. )
+    Local context as GameContext at me.contextPtr
+    Local currentTime as double = context.gameTime
+    
+    if me.oldTime = 0 Then 'First pass
+      me.tempsMonitor = 0
+      me.tempsStore = 0
+    Else
+      me.tempsMonitor += currentTime - me.oldTime
+      me.tempsStore += currentTime - me.oldTime
+    end if
+    
+    if me.tempsMonitor >= %DIFFICULTY_MONITOR_RATE Then
+      me.Rate
+      incr me.rateCount
+      me.tempsMonitor = 0
+    EndIf
+    
+    if me.tempsStore >= %DIFFICULTY_MONITOR_RATE Then
+      me.Store(currentTime)
+      me.tempsStore = 0
+      me.rateCount = 0
+    EndIf
+    
+    me.oldTime = currentTime
+    
+  end function
+
+
+  Function Store(currentTime as double)
+    if not me.rateCount > 0 then return
+    incr me.mesuresCount
+    Local toStore as monitorData
+    if me.mesuresCount > ubound(me.storage) Then
+      Decr me.mesuresCount
+      me.ExportToFile
+      me.mesuresCount = 1
+    endif
+    toStore.temps = currentTime - me.gamesessionStartTime
+    toStore.difficulty = me.oneData.difficulty / me.rateCount
+    toStore.meanCapacity = me.oneData.meanCapacity / me.rateCount
+    toStore.meanDistance = me.oneData.meanDistance / me.rateCount
+    toStore.meanFrameRate = me.oneData.meanframeRate / me.rateCount
+    
+    me.storage(me.mesuresCount) = toStore
+    
+    me.oneData.difficulty = 0
+    me.oneData.meanCapacity = 0
+    me.oneData.meanDistance =0
+    me.oneData.meanframeRate = 0
+    
+    '[t] Shame, following syntax does not work.
+    'me.storage(me.mesuresCount).temps = currentTime
+    'me.storage(me.mesuresCount).difficulty = me.difficulty / me.rateCount
```

Comment:
this syntax does not work, was it intended ?

Reply:
Sebastian,

the dynamic arrays of UDTs are not supported at all at the moment:
```
type Vec2D
  X as Single
  y as Single
end type

type Polygon
  points() as Vec2D
  
  function _Create()
    redim me.points(10)
    
    me.points(10).x = 1
  end function

end type

dim p as Polygon()
```

https://thinbasic.github.io/book-thinbasic_user_defined_types/chapter_01-01-01-dynamic-array-elements.html

Reply:
to be sure to understand this previous comment : the code is not valid because of `points() as Vec2D`,  right ?

Reply:
Exactly - because it is dynamic array of UDT. It should work in the future, Eros has it on the list.

I will ask him to cause runTime error in such a case, so user is not given hope it is working.